### PR TITLE
fix: don't render HTML Embed server-side in production if clientOnly option is used

### PR DESCRIPTION
**Pull request by kof**
## Code Review

Following tests are added
```
Published site
    ✓ clientOnly === false (30 ms)
    ✓ clientOnly === true (7 ms)
  Builder renderer= canvas | preview
    ✓ script processing are not applied when clientOnly=true executeScriptOnCanvas=false and renderer=canvas (4 ms)
    ✓ script processing are not applied when clientOnly=true executeScriptOnCanvas=undefined and renderer=canvas (2 ms)
    ✓ script processing are not applied when clientOnly=false executeScriptOnCanvas=false and renderer=canvas (4 ms)
    ✓ script processing are not applied when clientOnly=false executeScriptOnCanvas=undefined and renderer=canvas (5 ms)
    ✓ script processing are not applied when clientOnly=undefined executeScriptOnCanvas=false and renderer=canvas (7 ms)
    ✓ script processing are not applied when clientOnly=undefined executeScriptOnCanvas=undefined and renderer=canvas (2 ms)
    ✓ In preview mode, script processing are applied when clientOnly=true and renderer="preview" and executeScriptOnCanvas=true (3 ms)
    ✓ In preview mode, script processing are applied when clientOnly=true and renderer="preview" and executeScriptOnCanvas=false (2 ms)
    ✓ In preview mode, script processing are applied when clientOnly=true and renderer="preview" and executeScriptOnCanvas=undefined (2 ms)
    ✓ In preview mode, script processing are applied when clientOnly=false and renderer="preview" and executeScriptOnCanvas=true (2 ms)
    ✓ In preview mode, script processing are applied when clientOnly=false and renderer="preview" and executeScriptOnCanvas=false (2 ms)
    ✓ In preview mode, script processing are applied when clientOnly=false and renderer="preview" and executeScriptOnCanvas=undefined (2 ms)
    ✓ In preview mode, script processing are applied when clientOnly=undefined and renderer="preview" and executeScriptOnCanvas=true (2 ms)
    ✓ In preview mode, script processing are applied when clientOnly=undefined and renderer="preview" and executeScriptOnCanvas=false (2 ms)
    ✓ In preview mode, script processing are applied when clientOnly=undefined and renderer="preview" and executeScriptOnCanvas=undefined (2 ms)
    ✓ In preview mode, script processing are applied when clientOnly=true and renderer="canvas" and executeScriptOnCanvas=true (1 ms)
    ✓ In preview mode, script processing are applied when clientOnly=false and renderer="canvas" and executeScriptOnCanvas=true (3 ms)
    ✓ In preview mode, script processing are applied when clientOnly=undefined and renderer="canvas" and executeScriptOnCanvas=true (2 ms)
    ✓ Code change cause scripts to be updated (2 ms)
    ✓ Placeholder is shown if code is "" (1 ms)
    ✓ Placeholder is shown if code is "   "
```

- [ ] hi @istarkov , I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)
  - test it on preview

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio/blob/main/apps/builder/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `builder/env-check.js` if mandatory

### CODE REVIEW

File: packages/sdk-components-react/src/html-embed.test.tsx [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+});
+
+function cartesian<A, B>(a: A[], b: B[]): [A, B][];
+function cartesian<A, B, C>(a: A[], b: B[], c: C[]): [A, B, C][];
+
+// https://stackoverflow.com/questions/12303989/cartesian-product-of-multiple-arrays-in-javascript
+// eslint-disable-next-line func-style
+function cartesian(...a: unknown[][]) {
+  return a.reduce((a, b) => a.flatMap((d) => b.map((e) => [d, e].flat())));
+}
```
Comment (kof) [inline packages/sdk-components-react/src/html-embed.test.tsx:169]: this should be some utility in reusable place

File: packages/sdk-components-react/src/html-embed.tsx [hunk 2]
```diff
@@ -4,10 +4,15 @@ import {
...
   useRef,
   type ForwardedRef,
+  useSyncExternalStore,
+  useState,
+  type ReactNode,
 } from "react";
 import { mergeRefs } from "@react-aria/utils";
 import { ReactSdkContext } from "@webstudio-is/react-sdk";
 
+export const SCRIPT_PROCESSED_TEST_ID_PREFIX = "client-";
```
Comment (kof) [inline packages/sdk-components-react/src/html-embed.tsx:14]: I would like to standardize everytthing tests related being exported under the same name, e.g.

export const __testing__ = {
  scriptTestIdPrefix ....
}

We are often exporting things that are only useful for testing, so this convention would be practical

File: packages/sdk-components-react/src/html-embed.tsx [hunk 3]
```diff
@@ -83,16 +93,33 @@ const Placeholder = (props: ChildProps) => {
   );
 };
 
+const ClientOnly = (props: { children: ReactNode }) => {
+  const isServer = useSyncExternalStore(
```
Comment (kof) [inline packages/sdk-components-react/src/html-embed.tsx:97]: is it worth extracting it into a hook and commenting what this does with a link to that great article?

File: packages/sdk-components-react/src/html-embed.tsx [hunk 4]
```diff
@@ -158,26 +160,57 @@ type HtmlEmbedProps = {
...
     }
 
-    const execute = shouldExecute({
-      renderer,
-      executeScriptOnCanvas,
-      clientOnly,
-    });
+    if (ssrRendered) {
+      // We are on published site, on server rendering or after hydration
+
```
Comment (kof) [inline packages/sdk-components-react/src/html-embed.tsx:182]: unwanted newline

File: packages/sdk-components-react/src/html-embed.tsx [hunk 5]
```diff
@@ -158,26 +160,57 @@ type HtmlEmbedProps = {
...
+      );
     }
 
-    return <InnerHtml innerRef={ref} code={code} {...rest} />;
+    // Use key={code} to allow scripts to be reexecuted when code has changed
+    return (
+      <ClientOnly>
+        <ClientEmbed key={code} innerRef={ref} code={code} {...rest} />
+      </ClientOnly>
+    );
```
Comment (kof) [inline packages/sdk-components-react/src/html-embed.tsx:213]: please move this comment above key
```suggestion
    return (
      <ClientOnly>
        <ClientEmbed 
          // Use key={code} to allow scripts to be reexecuted when code has changed
          key={code} 
          innerRef={ref} code={code} {...rest} />
      </ClientOnly>
    );
```

File: packages/sdk-components-react/src/html-embed.test.tsx [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+// import ReactDOM from "react-dom";
+import ReactDOMServer from "react-dom/server";
+import { test, expect, describe } from "@jest/globals";
+// eslint-disable-next-line import/no-internal-modules
+import "@testing-library/jest-dom/jest-globals";
+import { fireEvent, render, screen } from "@testing-library/react";
+import { SCRIPT_PROCESSED_TEST_ID_PREFIX, HtmlEmbed } from "./html-embed";
+import { ReactSdkContext } from "@webstudio-is/react-sdk";
+
+global.React = React;
```
Comment (kof) [inline packages/sdk-components-react/src/html-embed.test.tsx:15]: that shouldn't be here, either lets add @todo to remove later with playwright rewrite or we need this centralized

**Review by istarkov — APPROVED** [review]

**Pull request merged**